Gem/Code/Source/ShaderReloadSoakTestComponent.cpp

```diff
@@ -0,0 +1,322 @@
+        AzFramework::StringFunc::Path::Join(m_tempSourceFolder.c_str(), newName, newFilePath);
+        if (io->Exists(newFilePath.c_str()))
+        {
+            if (!replaceIfExists)
+            {
+                return;
+            }
+        }
+
+        AZStd::string originalFilePath;
+        AzFramework::StringFunc::Path::Join(m_testDataFolder.c_str(), originalName, originalFilePath);
+        if (!io->Exists(originalFilePath.c_str()))
+        {
+            AZ_Error("ShaderReloadSoakTestComponent", false, "Could not find source file '%s'. This sample can only be used on dev platforms.", originalFilePath.c_str());
+            return;
+        }
+
+        // Instead of copying the file using AZ::IO::LocalFileIO, we load the file and write out a new file over top
+        // the destination. This is necessary to make the AP reliably detect the changes (if we just copy the file,
+        // sometimes it recognizes the OS level copy as an updated file and sometimes not).
+
+        AZ::IO::Path copyFrom = AZ::IO::Path(originalFilePath);
+        AZ::IO::Path copyTo = AZ::IO::Path(newFilePath);
+
+        m_fileIoErrorHandler.BusConnect();
+
+        auto readResult = AZ::Utils::ReadFile(copyFrom.c_str());
+        if (!readResult.IsSuccess())
+        {
+            m_fileIoErrorHandler.ReportLatestIOError(readResult.GetError());
+            return;
+        }
+
+        auto writeResult = AZ::Utils::WriteFile(readResult.GetValue(), copyTo.c_str());
+        if (!writeResult.IsSuccess())
+        {
+            m_fileIoErrorHandler.ReportLatestIOError(writeResult.GetError());
+            return;
+        }
+
+        m_fileIoErrorHandler.BusDisconnect();
+    }
+
+    void ShaderReloadSoakTestComponent::DeleteTestFile(const char* tempSourceFile)
+    {
+        AZ::IO::Path deletePath = AZ::IO::Path(m_tempSourceFolder).Append(tempSourceFile);
+
+        if (AZ::IO::LocalFileIO::GetInstance()->Exists(deletePath.c_str()))
+        {
+            m_fileIoErrorHandler.BusConnect();
+
+            if (!AZ::IO::LocalFileIO::GetInstance()->Remove(deletePath.c_str()))
+            {
+                m_fileIoErrorHandler.ReportLatestIOError(AZStd::string::format("Failed to delete '%s'.", deletePath.c_str()));
+            }
+
+            m_fileIoErrorHandler.BusDisconnect();
+        }
+    }
+
+    bool ShaderReloadSoakTestComponent::ReadInConfig(const AZ::ComponentConfig*)
+    {
+        return true;
+    }
+
+    void ShaderReloadSoakTestComponent::Activate()
+    {
+        AZ::TickBus::Handler::BusConnect();
+
+        InitTestDataFolders();
+        CopyTestFile(RedShaderFile, "Fullscreen.azsl");
+        CopyTestFile("Fullscreen.shader.txt", "Fullscreen.shader");
+        m_expectedPixelColor = RED_COLOR;
+
+        // A short wait is necessary as the pipeline creation will fatally fail
+        // if the "Fullscreen.azshader" doesn't exist.
+        AZStd::this_thread::sleep_for(AZStd::chrono::milliseconds(1000));
```

Comment:
Instead of sleeping, can you look for signals from the AP about the asset status? See the AssetStatusTracker class for some ideas.

Reply:
I was just being lazy. We'll try to use the AssetCollectionAsyncLoader or AssetStatusTracker

Reply:
+1